### DISCUSSION TITLE
Can I use the browser local cache with HTMX?

### DISCUSSION THREAD

**Discussion by hrstoyanov** [bigskysoftware/htmx#1531] [Q&A]
I am looking for some easy element  attributes to swap out chunks of the Dom with something from the local browser cache. Also , some more sophisticated scenarios:

1."Swap out the inner/outer html with chunk from local cache with key X, or fetch it from the server"

2. "Swap out the inner/outer html with cached chunk under key X but only it is older than Y milliseconds"

3. "Swap out ... From http request, but then also store the response in local cache"

Note:
1. This is different from server response cache mechanisms in http.

2. It is like an "htmx magic dust" sprinkled over the browsers cache api

**Comment by dz4k** [comment]
What advantage would this have over normal caching in browsers with the [Cache-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) header?

**Comment by binaryfire** [comment]
@dz4k I think the problem is cache invalidation. Let's say you're caching fragments of a dashboard for 5 minutes to prevent unnecessary requests and keep things snappy while a user is navigating around. And then one of those cached fragments needs to be invalidated because the user has added new data.

The only way (that I know of) for a server side webapp to clear the browser cache is using [Clear-Site-Data](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Clear-Site-Data). But that's not viable until it's supported in Safari. And knowing Apple that could be years away...

But if you implement caching in JS via local storage you can clear it whenever you want. Eg. when a form is submitted.

**Comment by HeyHugo** [comment]
@binaryfire / @hrstoyanov  You should check out this article on caching https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching
Section regarding validation may be of particular interest.

**Reply by binaryfire** [reply]
@HeyHugo I'm familiar with that article. But it doesn't change what I said above? Clear-Site-Data seems to be the only way a server can programmatically invalidate browser cache and it's not supported on Safari yet (unfortunately). And you'd still need to hit the server on every page load to check if the content has been modified.

Maybe I'm missing something... I'm coming from building dashboards with Tanstack Query where things like stats graphs, data tables etc. are cached client side until the user does something that needs them to be invalidated. At which point the the cache is purged from the store.

There's noticeable latency for users far away from our server (eg. Australia -> US) when the server needs to be hit on each page load to check for stale content. There probably isn't any way around it since state is server-side with HTMX. Which is fine: the hybrid approach of HTMX + a lightweight JS framework works well too.

**Comment by Nasjoe** [comment]
Something like an hx-cache with parameters that would allow the three scenarios mentioned by @hrstoyanov  in the first post would be great.

This could also open up possibilities for building web and mobile applications with an easily configurable offline mode.

Personally, I'm also very interested.

**Comment by nickchomey** [comment]
A service worker is probably the best option for implementing any of this (and opens up lots of other possibilities).

It will intercept all requests before going out to the network and returning to the dom. So you can store the responses in any of the browser storage apis - cache, local storage, session storage, indexeddb etc.

I'd suggest indexeddb as it offers the most possibilities. There's plenty of libraries made for working with all of this - Google workbox, idb, dexie, rxdb, etc

**Reply by Nasjoe** [reply]
I would sponsor this feature without hesitation!

**Reply by 7flash** [reply]
> I would sponsor this feature without hesitation!

Here's an example: [service_worker/index.tsx](https://github.com/galaxydo/icp-first-milestone-app/blob/master/src/galaxy-frontend/src/service_worker/index.tsx)

**Reply by nickchomey** [reply]
is there a non-minified version?

**Reply by 7flash** [reply]
> is there a non-minified version?

i've fixed my comment above, it's pointing to correct file now

it's a service worker which intercepts all requests from frontend to persist uploaded files in IndexedDB (there's no server at all in this example)

it's using lightning-fs (implements filesystem api) and wayne (alternative to express compatible with lightning-fs and service worker environment)

**Comment by AjaniBilby** [comment]
@hrstoyanov, I have experimented at my work with doing pre-renders in htmx.
So it does a swap with a pre-loaded results, then does the request against the stale html with fresh data.

I would warn it gets messy very quickly
> i.e. if the user is interacting with an element/form/field from the stale data, then the fresh gets swapped much later due to a latency spike, and now user inputs are interrupted or lost

Potentially you could adapt what I've done in the past for your own needs, everything is MIT licensed so do with it what you will.
[https://github.com/AjaniBilby/hx-prep](Ajanibilby/hx-prep).
Luckily for me, my work is on an internal intranet website, so I'm not dealing with international networking latency, so the answer for me was:
  - Smarter lazy loading to get inital page loads
  - Use [ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/ETag)s to ensure lazy loads weren't recomputed if the data hadn't changed (some JS for easy server side ETag validation and handling [source](https://github.com/AjaniBilby/htmx-router/blob/main/source/response.ts#L78-L158))
  - Better database structure, query optimizations and indexing

I highly recommend learning about eTags if you don't already know about them and your latecy is coming from resolution time rather than the network. Some times I am lazy enough that I make my tag be just the unix timestamp rounded down to the nearest 15 minute internal and radix 36 encoded. That way all users are seeing the same changed versions rather than potentially out of sync `Cache-Control` ages
